shared/modules/caip-stream.ts

```diff
@@ -0,0 +1,81 @@
+import { isObject } from '@metamask/utils';
+import { Transform, pipeline, Duplex } from 'readable-stream';
+
+export class SplitStream extends Duplex {
+  substream: Duplex;
+
+  constructor(substream?: SplitStream) {
+    super({ objectMode: true });
+    this.substream = substream ?? new SplitStream(this);
+  }
+
+  _read() {
+    return undefined;
+  }
+
+  _write(
+    value: unknown,
+    _encoding: BufferEncoding,
+    callback: (error?: Error | null) => void,
+  ) {
+    this.substream.push(value);
+    callback();
+  }
+}
```

Comment:
similar to https://github.com/MetaMask/core/blob/main/packages/json-rpc-middleware-stream/src/createEngineStream.ts#L16